tamr_toolbox/dataset/manage.py

```diff
@@ -0,0 +1,527 @@
+    """
+
+    if not dataset and not primary_keys:
+        raise ValueError(f"dataset or primary_keys must be defined")
+
+    # Get dataset information
+    if dataset:
+        # Get attributes from dataset object
+        attribute_types, attribute_descriptions = {}, {}
+        for attr in dataset.attributes.stream():
+            attribute_types[attr.name] = attribute_type.from_json(attr.type.spec().to_dict())
+            attribute_descriptions[attr.name] = attr.description
+        attributes = attribute_types.keys()
+        # Get dataset spec information
+        description = dataset.description
+        tags = dataset.tags
+        primary_keys = dataset.key_attribute_names
+
+    # Check input type is correct
+    if attributes and not isinstance(attributes, Iterable):
+        raise TypeError("attributes arg must be an Iterable")
+
+    dataset_exists = exists(client=client, dataset_name=dataset_name)
+    if not dataset_exists:
+        creation_spec = {
+            "name": dataset_name,
+            "description": description,
+            "keyAttributeNames": primary_keys,
+            "externalId": external_id,
+            "tags": tags,
+        }
+        client.datasets.create(creation_spec)
+        LOGGER.info(f"A dataset with name {dataset_name} has been created")
+    else:
+        raise ValueError(f"A dataset with name '{dataset_name}' already exists")
+
+    # Get new dataset
+    target_dataset = client.datasets.by_name(dataset_name)
+
+    # Update attributes in dataset
+    if attributes:
+        filtered_attributes = [attr for attr in attributes if attr not in primary_keys]
+        create_attributes(
+            dataset=target_dataset,
+            attributes=filtered_attributes,
+            attribute_types=attribute_types,
+            attribute_descriptions=attribute_descriptions,
+        )
+
+    return target_dataset
+
+
+def update(
+    dataset: Dataset,
+    *,
+    attributes: Optional[Iterable[str]] = None,
+    attribute_types: Optional[Dict[str, attribute_type.AttributeType]] = None,
+    attribute_descriptions: Optional[Dict[str, str]] = None,
+    description: Optional[str] = None,
+    tags: Optional[List[str]] = None,
+    override_existing_types: bool = False,
+) -> Dataset:
+    """
+    Flexibly update a source dataset in Tamr. All the attributes that should exist in the dataset
+    must be defined in the attributes argument. This function will add/remove attributes in the
+    dataset until the dataset attributes matches the set of attributes passed in as an argument.
+    The default attribute type will be ARRAY STRING. To set non-default attribute types, you must
+    define them in the attribute_types dictionary. Any attribute descriptions can be specified in
+    the attribute_descriptions dictionary. By default, the existing attribute types will not
+    change unless override_existing_types is set to True. When it's False, the attribute type
+    updates will only be logged.
```

Comment:
nit: format and consistent grammar

```suggestion
    """Flexibly update a source dataset in Tamr

    All the attributes that should exist in the dataset must be defined in the attributes argument.
    This function will add/remove attributes in the dataset until the dataset attributes matches
    the set of attributes passed in as an argument. The default attribute type will be ARRAY
    STRING . To set non-default attribute types, they must be defined in the attribute_types
    dictionary. Any attribute descriptions can be specified in the attribute_descriptions
    dictionary. By default, the existing attribute types will not change unless
    override_existing_types is set to True. When False, the attribute type updates will only be
    logged.
```